src/main/java/io/fabric8/maven/docker/access/ecr/AwsSigner4Request.java

```diff
@@ -0,0 +1,223 @@
+        canonicalHeaders = canonical.toString();
+        signedHeaders = signed.toString();
+    }
+
+    public String getRegion() {
+        return region;
+    }
+
+    public String getService() {
+        return service;
+    }
+
+    public String getSigningDate() {
+        return signingDate;
+    }
+
+    public String getSigningDateTime() {
+        return signingDateTime;
+    }
+
+    public String getScope() {
+        return scope;
+    }
+
+    public String getMethod() {
+        return method;
+    }
+
+    public URI getUri() {
+        return uri;
+    }
+
+    public String getCanonicalHeaders() {
+        return canonicalHeaders;
+    }
+
+    public String getSignedHeaders() {
+        return signedHeaders;
+    }
+
+    private static String getSigningDateTime(HttpRequest request, Date signingTime) {
+        Header dateHeader = request.getFirstHeader("X-Amz-Date");
+        if(dateHeader!=null) {
+            return dateHeader.getValue();
+        }
+        synchronized (TIME_FORMAT) {
+            return TIME_FORMAT.format(signingTime);
+        }
+    }
+
+    private static URI getUri(HttpRequest request) {
+        String hostName = request.getFirstHeader("host").getValue();
+        String requestTarget = request.getRequestLine().getUri();
+        URI requestUri = createUri(hostName, requestTarget);
+        return requestUri.normalize();
+    }
+
+    private static URI createUri(String authority, String uri) {
+        String scheme = "https";
+        int schemeEnd = uri.indexOf(':');
+        int pathStart = uri.indexOf('/');
+        if(schemeEnd>=0 && schemeEnd < pathStart) {
+            scheme = uri.substring(0, schemeEnd);
+            if(uri.charAt(pathStart + 1) == '/') {
+                authority = uri.substring(pathStart);
+                pathStart = uri.indexOf('/', pathStart+2);
+            }
+        }
+
+        String path;
+        String query;
+        int queryIdx = uri.indexOf('?', pathStart);
+        if(queryIdx<0) {
+            query =  null;
+            path = uri.substring(pathStart);
+        }
+        else {
+            query = uri.substring(queryIdx+1);
+            path = uri.substring(pathStart, queryIdx);
+        }
+        try {
+            return new URI(scheme, authority, path, query, null);
+        }
+        catch(URISyntaxException e) {
+            throw new UndeclaredThrowableException(e);
+        }
+    }
+
+    private static void canonicalizeHeaders(HttpRequest request, StringBuilder canonical, StringBuilder signed) {
+        Map<String, StringBuilder> unique = new TreeMap<>();
+        for (Header header : request.getAllHeaders()) {
+            String key = header.getName().toLowerCase(Locale.US);
+            if (key.equals("connection")) {
+                continue;
+            }
+            StringBuilder builder = unique.get(key);
+            if (builder != null) {
+                if (builder.length() > 0) {
```

Comment:
refactored this method.  complicated by fact it was trying to accomplish two tasks in same loop.